File: my-submissions/1046-max-consecutive-ones-iii/solution.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < nums.length; right++) {
             // If we see a 0, consume one allowed flip
            if (nums[right] == 0) {
                k--;
            }

            // out of flips, shift left side in once
            if (k < 0) {
                if (nums[left] == 0) {
                    k++;
                }
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }
}

File: my-submissions/1046-max-consecutive-ones-iii/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function (nums, k) {
    let left = 0;
    let maxLength = 0;

    // slide to right
    for (let right = 0; right < nums.length; right++) {
        // If we encounter a 0, we decrease our allowed flips (k)
        if (nums[right] === 0) {
            k--;
        }

        // If k < 0, the current window is invalid. 
        // We shift the left pointer by 1 to maintain the window size.
        if (k < 0) {
            if (nums[left] === 0) {
                // We give back a flip if a 0 leaves the window
                k++;
            }
            left++;
        }
         
        // current window size => right - left + 1
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
};
